### PR TITLE
Improves reduction code and removes `kernels/boolean_reductions.hpp`

### PR DESCRIPTION
**Pull request by ndgrigorian**
This pull request further modularizes reduction code and fully moves boolean reductions (any and all) functionality into `reductions.hpp`.

Sequential reductions now have a `submit_sequential_reduction` wrapper, reducing repetitious kernel calls.

Boolean reductions now use the same basic kernels as other reductions.

Additionally, reduction type dispatching has been moved into individual reduction files, i.e., `reductions/max.cpp` now contains the dispatch factories, and so on for all reductions. This significantly reduces the amount of code in `kernels/reductions.hpp`.

- [X] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

File: dpctl/tensor/libtensor/include/kernels/reductions.hpp [hunk 1]
```diff
@@ -318,28 +379,244 @@ struct CustomReductionOverGroupWithAtomicFunctor
...
+            else if constexpr (std::is_same_v<ReductionOp, sycl::maximum<outT>>)
+            {
+                res_ref.fetch_max(red_val_over_wg);
+            }
+            else if constexpr (std::is_same_v<ReductionOp, sycl::minimum<outT>>)
+            {
+                res_ref.fetch_min(red_val_over_wg);
+            }
+            else if constexpr (std::is_same_v<ReductionOp,
+                                              sycl::logical_and<outT>>) {
```
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/include/kernels/reductions.hpp:394]: Should we use `su_ns::IsLogicalAnd<outT>::value` instead?

File: dpctl/tensor/libtensor/include/kernels/reductions.hpp [hunk 2]
```diff
@@ -111,7 +123,16 @@ struct SequentialReduction
             const ssize_t inp_offset = inp_iter_offset + inp_reduction_offset;
 
             using dpctl::tensor::type_utils::convert_impl;
-            outT val = convert_impl<outT, argT>(inp_[inp_offset]);
+            outT val;
+            if constexpr (std::is_same_v<ReductionOp,
+                                         sycl::logical_and<outT>> ||
+                          std::is_same_v<ReductionOp, sycl::logical_or<outT>>)
```
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/include/kernels/reductions.hpp:129]: Use `su_ns::IsLogicalOr<outT, ReductionOp>::value || su_ns::IsLogicalAnd<outT, ReductionOp>::value`.

Comment (oleksandr-pavlyk) [reply dpctl/tensor/libtensor/include/kernels/reductions.hpp:129]: I think this same remark applies to other instances in this file.

**Comment by oleksandr-pavlyk** [comment]
New test failure:

```
_________________________ test_max_min_nan_propagation _________________________

    def test_max_min_nan_propagation():
        get_queue_or_skip()
    
        # float, finites
        x = dpt.arange(4, dtype="f4")
        x[0] = dpt.nan
>       assert dpt.isnan(dpt.max(x))
E       AssertionError: assert usm_ndarray(False)
E        +  where usm_ndarray(False) = <UnaryElementwiseFunc 'isnan'>(usm_ndarray(-inf, dtype=float32))
E        +    where <UnaryElementwiseFunc 'isnan'> = dpt.isnan
E        +    and   usm_ndarray(-inf, dtype=float32) = <function max at 0x7fbe36c69580>(usm_ndarray([nan,  1.,  2.,  3.], dtype=float32))
E        +      where <function max at 0x7fbe36c69580> = dpt.max

/usr/share/miniconda/envs/test_dpctl/lib/python3.11/site-packages/dpctl/tests/test_usm_ndarray_reductions.py:140: AssertionError
```

**Comment by ndgrigorian** [comment]
> New test failure:
> 
> ```
> _________________________ test_max_min_nan_propagation _________________________
> 
>     def test_max_min_nan_propagation():
>         get_queue_or_skip()
>     
>         # float, finites
>         x = dpt.arange(4, dtype="f4")
>         x[0] = dpt.nan
> >       assert dpt.isnan(dpt.max(x))
> E       AssertionError: assert usm_ndarray(False)
> E        +  where usm_ndarray(False) = <UnaryElementwiseFunc 'isnan'>(usm_ndarray(-inf, dtype=float32))
> E        +    where <UnaryElementwiseFunc 'isnan'> = dpt.isnan
> E        +    and   usm_ndarray(-inf, dtype=float32) = <function max at 0x7fbe36c69580>(usm_ndarray([nan,  1.,  2.,  3.], dtype=float32))
> E        +      where <function max at 0x7fbe36c69580> = dpt.max
> 
> /usr/share/miniconda/envs/test_dpctl/lib/python3.11/site-packages/dpctl/tests/test_usm_ndarray_reductions.py:140: AssertionError
> ```

This was caused by the atomic `fetch_min/max` operations being called for reductions which used our custom implementations.

To fix this, I added more utilities for checking if an operator is specifically a sycl op. That way if a work-around is needed for `reduce_over_group` in the future, but we are still using the sycl op, we can leverage atomics better.

**Review by oleksandr-pavlyk — APPROVED** [review]
LGMT! Thank you @ndgrigorian !

**Pull request merged**